### PR TITLE
feat(gamma-apim): integrate policy studio page with V2 compatibility

### PR DESCRIPTION
**Pull request by jourdiw**
### Context sidebar adoption

Migrates both `ApiDetailLayout` and `ApiProductDetailLayout` from hand-built `<aside>`/`<main>` flex layouts to graphene-core's `ContextSidebar` + `useLayoutConfig` pattern. The shell (`ShellLayout`, `LocalDevShell`) is updated to pass through the new layout slot properties (`viewMode`, `contextExpanded`, `contextSidebar`, `contentVariant`, `leading`).

**API detail (`ApiDetailLayout`):**
- Uses `ContextSidebar` with `ApiInfoHeader` (avatar, name, state badge, description, tech chips) as the header and `ApiDetailSidebarNav` as the navigation body.
- `ContextToggleButton` in the `leading` slot allows collapsing/expanding the sidebar.
- Breadcrumbs: `API Proxies > {api name}`.

**API Products (`ApiProductDetailLayout`):**
- Same pattern: `ContextSidebar` with `ProductInfoHeader` and `ApiProductSidebarNav`.
- Removes the old `<aside>` + manual divider + `<main>` with hardcoded `height: calc(100dvh - 5rem)`.
- Breadcrumbs: `API Products > {product name}`.

### Deploy banner redesign

Replaces the bulky padded `Alert`-based deploy banner with a compact single-line status strip (~40px vs ~80px):
- Uses a `border-b` flex row with a subtle `color-mix(in oklab, var(--color-warning) 6%, transparent)` background tint.
- Text uses `text-foreground` for contrast (WCAG-compliant), with `text-warning` on the icon only.
- `TriangleAlertIcon` replaces `MessageSquareWarningIcon`.
- Copy changed from "Pending changes / This API is out of sync" to "This API has undeployed changes."
- Removed unused `Alert`, `AlertAction`, `AlertDescription`, `AlertTitle` imports.
- Test assertions updated to match the new copy; unused Alert mocks removed from the spec file.
- Filed [gravitee-ui-graphene#346](https://github.com/gravitee-io/gravitee-ui-graphene/issues/346) to add a `banner` layout slot to `AppLayout` for full-width rendering in a future follow-up.

Screenshot 2026-05-28 at 20 43 42

### Policy Studio integration

Replaces the Policy Studio placeholder route with a full integration of `@gravitee/graphene-policy-studio`:
- Service layer, React Query hooks, type definitions, and a V2-to-V4 flow adapter for backward compatibility.
- V2 APIs render read-only; V4/native APIs support full editing.
- Uses `contentVariant: 'full-bleed'` for edge-to-edge rendering.

Screenshot 2026-05-28 at 20 43 55

### Shared API client

Adds `apimFetchJsonV2Org` to the shared APIM client for org-scoped v2 endpoint calls required by the policy/plugin catalog.

### Merge fix

Resolves duplicate JSX props (`viewMode`, `contextExpanded`, `contextSidebar`, `contentVariant`, `leading`) left by a merge conflict in `ShellLayout` and `LocalDevShell`.

## Notes

- Error handling/toast feedback in the save hook (`usePolicyStudioSave`) is deferred to a follow-up.
- The deploy banner renders inside the content padding area for now. Full-width rendering will land once the `banner` slot is added to graphene-core ([#346](https://github.com/gravitee-io/gravitee-ui-graphene/issues/346)).

## Test plan

- [x] Navigate to an existing V4 API -> Policy Studio — verify the studio loads with correct flows, plans, policies, and connectors
- [x] Edit a flow and save — verify the GET-then-PUT save roundtrip works and the UI reflects the saved state
- [x] Navigate to a V2 API -> Policy Studio — verify the studio renders in read-only mode with V2 flows adapted to V4 shape
- [x] Collapse/expand the context sidebar via the toggle button on both API detail and API Product detail — verify layout behavior
- [x] Verify the deploy banner appears when an API has `deploymentState: NEED_REDEPLOY` and the user has `api-definition-u` permission
- [x] Click "Deploy API" on the banner — verify the deploy mutation fires and the banner disappears after success
- [x] Navigate to an API Product detail — verify the context sidebar renders with product info header, sync status badge, and nav links
- [x] Verify plugin catalog (policies, entrypoints, endpoints, shared policy groups) loads correctly

### CODE REVIEW

File: gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/pages/policy-studio/PolicyStudioPage.tsx [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    );
+
+    if (studioData.isLoading) {
+        return (
+            <div className="flex flex-col gap-4 p-6">
+                <Skeleton className="h-8 w-48 rounded" />
+                <Skeleton className="h-64 w-full rounded" />
+            </div>
+        );
+    }
```
Comment (jourdiw) [inline gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/pages/policy-studio/PolicyStudioPage.tsx]: The studio handles loading itself so this can be removed

**Review by jourdiw — COMMENTED** [review]
Not all of the services are strict to the policy studio as well as the types. They should all be added rather to the services/ folder.

The types should then as well be in the types/ folder, not under the feature unless they are specific to the UI.

File: gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/hooks/usePolicyStudioSave.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+            const envId = env.id;
+
+            const { commonFlows, flowExecution, plansToUpdate } = output;
+
+            if (commonFlows || flowExecution) {
+                const current = await getFullApiDetail(envId, apiId);
+                const updated: Record<string, unknown> = { ...current };
+                if (commonFlows) updated.flows = commonFlows;
+                if (flowExecution) updated.flowExecution = flowExecution;
+                await updateApi(envId, apiId, updated);
```
Comment (jourdiw) [inline gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/hooks/usePolicyStudioSave.ts]: **[Score: 72/100]** ⚠️ The `onSave` callback has no error handling. If `updateApi` or `getAndUpdatePlanFlows` rejects (network error, 409 conflict, validation failure), the promise propagates to `PolicyStudio`'s `onSave` — but it's unclear whether that component surfaces the error to the user or swallows it. At minimum, consider wrapping in a try/catch that shows a toast or alert on failure:

```ts
try {
    // ... existing logic
} catch (err) {
    // Surface error to user (toast, alert, etc.)
    throw err; // re-throw so PolicyStudio knows save failed
}
```

Also: spreading the full API detail (`{ ...current }`) and PUTting it back could send computed/read-only fields (`_links`, `definitionContext`, etc.) that the backend might reject or ignore. Consider filtering to only writable fields before the PUT.

File: gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/utils/v2FlowAdapter.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    };
+}
+
+/**
+ * Converts a V2 flow to the V4 Flow shape expected by `@gravitee/graphene-policy-studio`.
+ *
+ * V2 uses `pathOperator` + `methods` + `pre`/`post`.
+ * V4 uses `selectors` + `request`/`response`.
+ */
+export function mapFlowV2ToV4(flow: FlowV2): Flow {
```
Comment (jourdiw) [inline gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/utils/v2FlowAdapter.ts:40]: **[Score: 75/100]** ⚠️ The V2→V4 flow adapter is non-trivial mapping logic (path operators, condition selectors, step conversion) with **zero test coverage**. This is the kind of code that benefits most from unit tests because:

1. V2 APIs can have edge cases (empty `pathOperator`, flows with `pre` but no `post`, disabled flows).
2. The adapter will be exercised in read-only mode for legacy APIs — bugs here silently corrupt the visual display.
3. The `as Selector` type assertions mask potential shape mismatches.

A small `v2FlowAdapter.test.ts` with 5-6 cases (empty flow, single step, multi-step with condition, no pathOperator) would provide high confidence cheaply.

File: gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/components/detail/ApiDetailLayout.tsx [hunk 4]
```diff
@@ -187,6 +183,24 @@ export function ApiDetailLayout() {
...
+            viewMode: 'context',
+            contextExpanded,
+            contextSidebar: (
+                <ContextSidebar header={<ApiInfoHeader api={api ?? null} isLoading={isLoading} />}>
+                    <ApiDetailSidebarNav groups={API_PROXY_NAV_GROUPS} basePath={basePath} permissionsReady={permissionsReady} />
+                </ContextSidebar>
+            ),
+            leading: <ContextToggleButton expanded={contextExpanded} onToggle={() => setContextExpanded(v => !v)} />,
+            breadcrumbs: [
+                { label: 'API Proxies', href: `${basePath.slice(0, basePath.lastIndexOf('/apis/'))}${'/apis'}` },
```
Comment (jourdiw) [inline gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/components/detail/ApiDetailLayout.tsx:197]: **[Score: 55/100]** 💡 Breadcrumb `href` is built via `basePath.slice(0, basePath.lastIndexOf('/apis/'))` — this is fragile string surgery that silently breaks if the route structure ever changes (e.g. renamed segment, extra nesting). Same pattern in `ApiProductDetailLayout.tsx`.

Consider deriving it from the router's own path resolution (React Router's `useResolvedPath('..')` or the already-available `rootPath` from `useModuleRouting`) rather than manual string slicing. Alternatively, pass the list-page href down via context or a config constant.

File: gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/services/policyStudioService.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+export async function getFullApiDetail(envId: string, apiId: string): Promise<PolicyStudioApiDetail> {
+    return apimFetchJsonV2<PolicyStudioApiDetail>(envId, `/apis/${encodeURIComponent(apiId)}`);
+}
+
+export async function listPublishedPlans(envId: string, apiId: string): Promise<PolicyStudioPlanPage> {
+    return apimFetchJsonV2<PolicyStudioPlanPage>(
+        envId,
+        `/apis/${encodeURIComponent(apiId)}/plans?statuses=PUBLISHED,DEPRECATED&page=1&perPage=9999`,
```
Comment (jourdiw) [inline gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/services/policyStudioService.ts:64]: **[Score: 30/100]** 🎨 `perPage=9999` is a pragmatic escape hatch for loading all published plans at once. Worth a code comment documenting why the policy studio needs all plans upfront (to render all plan flows in the visual editor) so future readers don't try to "fix" it with proper pagination.

File: gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/hooks/usePolicyStudioData.ts [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+}
+
+export function usePolicyStudioData(apiId: string | undefined): PolicyStudioData {
+    const env = useEnvironment();
+    const envId = env?.id ?? '';
+    const enabled = Boolean(env && apiId);
+
+    const apiQuery = useQuery({
+        queryKey: policyStudioKeys.api(envId, apiId ?? ''),
+        queryFn: () => getFullApiDetail(envId, apiId!),
```
Comment (jourdiw) [inline gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/hooks/usePolicyStudioData.ts:104]: **[Score: 35/100]** 🎨 The `apiQuery` uses `queryFn: () => getFullApiDetail(envId, apiId!)` with a non-null assertion. The `enabled` guard ensures `apiId` is truthy before the query fires, so this is safe at runtime — but the assertion hides a type constraint from the compiler. Consider narrowing the type in the `enabled` check and using the conditional `select` pattern, or simply providing a fallback:

```ts
queryFn: () => getFullApiDetail(envId, apiId ?? ''),
```

(The function already `encodeURIComponent`s the id, so an empty string would produce a 404 rather than a crash — though the `enabled` flag prevents this case anyway.)

File: gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/utils/v2FlowAdapter.ts (hunk 3 above)
Comment (jourdiw) [reply gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/utils/v2FlowAdapter.ts:40]: Yes implement test cases

File: gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/services/policyStudioService.ts (hunk 5 above)
Comment (jourdiw) [reply gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/services/policyStudioService.ts:64]: We are obliged to do this for now.

File: gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/pages/policy-studio/PolicyStudioPage.tsx (hunk 1 above)
Comment (jourdiw) [reply gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/pages/policy-studio/PolicyStudioPage.tsx]: Done — removed the page-level loading skeleton. `PolicyStudio` handles its own loading state internally.

File: gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/hooks/usePolicyStudioSave.ts (hunk 2 above)
Comment (jourdiw) [reply gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/hooks/usePolicyStudioSave.ts]: Added try/catch that logs and re-throws so PolicyStudio still receives the rejection. Did not add toast yet (no shared toast pattern in this module). Skipped filtering writable fields — GET-then-PUT with full body matches existing module pattern (`getAndUpdatePlanFlows` does the same) and the backend ignores read-only fields.

File: gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/utils/v2FlowAdapter.ts (hunk 3 above)
Comment (jourdiw) [reply gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/utils/v2FlowAdapter.ts:40]: Done — added `v2FlowAdapter.spec.ts` with 14 test cases covering pathOperator, condition, combined selectors, empty pre/post, disabled flows, mapFlowsV2ToV4, mapFlowModeToExecution, and the default connector constants.

File: gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/components/detail/ApiDetailLayout.tsx (hunk 4 above)
Comment (jourdiw) [reply gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/components/detail/ApiDetailLayout.tsx:197]: Agree it's fragile, but `useDetailBasePath` already uses the same pathname-slicing approach and `ApiProductDetailLayout` uses the identical breadcrumb pattern. Refactoring to `useResolvedPath` or module routing would be a separate cross-layout change — happy to track as follow-up.

File: gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/services/policyStudioService.ts (hunk 5 above)
Comment (jourdiw) [reply gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/services/policyStudioService.ts:64]: Added an inline comment explaining Policy Studio needs all published plans in one request. `perPage=9999` is intentional until the studio supports paginated plan loading.

File: gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/hooks/usePolicyStudioData.ts (hunk 6 above)
Comment (jourdiw) [reply gravitee-gamma/gravitee-gamma-module-apim/src/main/ui/features/apis/hooks/usePolicyStudioData.ts:104]: Keeping `apiId!` — the `enabled` guard guarantees it is defined before `queryFn` runs. This is the standard TanStack Query pattern documented in their guides. `apiId ?? ''` would be a misleading fallback that never actually executes.

**Comment by jourdiw** [comment]
Re: services/types placement — they are already under `features/apis/services/` and `features/apis/types/`, consistent with other feature services (`plans.ts`, `apis.ts`, etc.). Moving the shared-capable calls (e.g. `listPolicies`, `listEntrypointPlugins`) to `shared/` is a valid follow-up but out of scope for this integration PR. Will track separately.

**Review by Okhelifi — APPROVED** [review]
